### PR TITLE
Use virtio console when possible to speedup extratests

### PR DESCRIPTION
**Pull request by dzedro**
- Related ticket: #10137 
  Fixed serial_terminal on passwordles login on LIVECD, changed passwordless setup in
  consoltest_setup on opensuse without bernhard user (get_var('ROOTONLY') || get_var('LIVECD'))
  and sshd is not using passwordless setup done in consoltest_setup.
- Verification run:
https://openqa.suse.de/tests/4257161 15-SP1 extratests
https://openqa.suse.de/tests/4256927 12-SP5 s390x
15-SP2
https://openqa.suse.de/tests/4256216 staging gnome
https://openqa.suse.de/tests/4256217 staging minimal-base
https://openqa.suse.de/tests/4256219 extra_tests_textmode x86_64
https://openqa.suse.de/tests/4256220 extra_tests_textmode s390x
https://openqa.suse.de/tests/4256221 extra_tests_textmode ppc64le
https://openqa.suse.de/tests/4257148 extra_tests_textmode aarch64
TW
https://openqa.opensuse.org/tests/1272652 gnome_live i686
https://openqa.opensuse.org/tests/1272653 gnome_live x86_64
https://openqa.opensuse.org/tests/1272654 jeos_extra x86_64
https://openqa.opensuse.org/tests/1272860 extra_tests_in_textmode x86_64
https://openqa.opensuse.org/tests/1272656 gnome x86_64
https://openqa.opensuse.org/tests/1272657 kde x86_64
https://openqa.opensuse.org/tests/1272658 rootonly x86_64
https://openqa.opensuse.org/tests/1272659 kde-live x86_64
https://openqa.opensuse.org/tests/1272660 xfce-live x86_64
https://openqa.opensuse.org/tests/1271322 jeos-extra aarch64
There is big queue on aarch64 and nothing changed for aarch64 since previous version
15.1
https://openqa.opensuse.org/tests/1273585 krypton-live

### CODE REVIEW

File: lib/serial_terminal.pm [hunk 1]
```diff
@@ -96,6 +96,10 @@ sub login {
         type_string("\n");
         die 'Failed to confirm that login was successful' unless wait_serial(qr/$escseq* \w+:~\s\# $escseq* \s*$/x);
     }
+    else {
+        # wait for shell to avoid PS1 failures on passwordless login e.g. LIVECD
+        die 'Failed to confirm that passwordless login was successful' unless wait_serial(qr/.*:~\s\#/);
```
Comment (Vogtinator) [inline lib/serial_terminal.pm:101]: AFAICT this also matches for the case above, so just move that out?

File: tests/console/ant.pm [hunk 2]
```diff
@@ -91,8 +91,8 @@ sub run {
     assert_script_run "cd $dir";
     assert_script_run "mkdir -p src/test";
     assert_script_run "mkdir -p build/classes";
-    assert_script_run "echo '$java_hello_world' >> src/test/Hello.java";
-    assert_script_run "echo '$build_file_xml' >> build.xml";
+    script_output("echo '$java_hello_world' >> src/test/Hello.java");
```
Comment (Vogtinator) [inline tests/console/ant.pm:94]: What's the purpose of this change?

File: tests/console/consoletest_setup.pm [hunk 3]
```diff
@@ -11,47 +11,64 @@
...
 sub run {
     my $self = shift;
-    # let's see how it looks at the beginning
-    save_screenshot;
-    check_var("BACKEND", "ipmi") ? use_ssh_serial_console : select_console 'root-console';
+    $self->select_serial_terminal;
+
+    systemctl('start sshd');
+
+    # generate ssh key and use same key for root and bernhard
```
Comment (Vogtinator) [inline tests/console/consoletest_setup.pm:38]: `bernhard` should not be hardcoded anywhere - it's `testapi::$username`.

File: tests/console/consoletest_setup.pm [hunk 4]
```diff
@@ -11,47 +11,64 @@
...
+    if (script_run('! test -e ~/.ssh/id_rsa') == 0) {
+        assert_script_run('ssh-keygen -t rsa -N "" -f ~/.ssh/id_rsa');
+    }
+
+    # copy and add root key into authorized_keys and public key into known_hosts of both root and bernhard
+    if (get_var('ROOTONLY') || get_var('LIVECD')) {
```
Comment (Vogtinator) [inline tests/console/consoletest_setup.pm:44]: For LIVECD, `testapi::$username` is `linux`.

File: tests/console/gdb.pm [hunk 5]
```diff
@@ -39,18 +36,15 @@ sub wait_serial_or_die {
 
 
 sub run {
-    #Setup console for text feedback.
-    select_console("root-console");
-    if (is_sle('=12-SP5') && is_aarch64()) {
```
Comment (Vogtinator) [inline tests/console/gdb.pm:44]: Where did this go?

File: tests/installation/bootloader_uefi.pm [hunk 6]
```diff
@@ -57,6 +57,11 @@ sub run {
         return;
     }
 
+    # press key up to cancel 10 sec timeout on jeos images to don't miss the grub screen
```
Comment (Vogtinator) [inline tests/installation/bootloader_uefi.pm:60]: Unrelated change?

File: tests/x11/exiv2.pm [hunk 7]
```diff
@@ -96,13 +96,15 @@ sub run {
     script_run "eog 20190201_154421.jpg", 0;
     assert_screen "exiv2_rename_test";
     send_key 'alt-f4';
+    wait_still_screen(1);
```
Comment (Vogtinator) [inline tests/x11/exiv2.pm:99]: Unrelated change?

File: tests/console/consoletest_setup.pm (hunk 3 above)
Comment (dzedro) [reply tests/console/consoletest_setup.pm:38]: sure, I will change that

File: tests/console/gdb.pm (hunk 5 above)
Comment (dzedro) [reply tests/console/gdb.pm:44]: This was some leftover from past, not needed at all.

Comment (Vogtinator) [reply tests/console/gdb.pm:44]: Should be part of a separate commit still

File: tests/installation/bootloader_uefi.pm (hunk 6 above)
Comment (dzedro) [reply tests/installation/bootloader_uefi.pm:60]: This is failure which were blocking VR, it has separate commit.
I didn't see any poo so i just added it to this PR.

File: tests/x11/exiv2.pm (hunk 7 above)
Comment (dzedro) [reply tests/x11/exiv2.pm:99]: This is failure which were blocking VR, it has separate commit.
I didn't see any poo so i just added it to this PR.

File: tests/console/consoletest_setup.pm (hunk 4 above)
Comment (dzedro) [reply tests/console/consoletest_setup.pm:44]: Will change hardcoded user to `testapi::$username`

File: lib/serial_terminal.pm (hunk 1 above)
Comment (dzedro) [reply lib/serial_terminal.pm:101]: This is fixing this failure when no password is needed, without waiting for shell it will immediately after login type to set PS1 which will fail as the shell is not ready.

https://openqa.opensuse.org/tests/1263528#step/consoletest_setup/3
```
[2020-05-12T05:57:33.206 CEST] [debug] >>> testapi::wait_serial: (?^ui:login:\s*$): ok
[2020-05-12T05:57:33.206 CEST] [debug] tests/console/consoletest_setup.pm:34 called opensusebasetest::select_serial_terminal -> lib/opensusebasetest.pm:1113 called testapi::select_console -> lib/susedistribution.pm:746 called serial_terminal::login -> lib/serial_terminal.pm:92 called testapi::type_string
[2020-05-12T05:57:33.206 CEST] [debug] <<< testapi::type_string(text="root\n")
[2020-05-12T05:57:33.206 CEST] [debug] <<< consoles::serial_screen::type_string(json_cmd_token="KfkNBVNu", text="root\n", cmd="backend_type_string")
[2020-05-12T05:57:33.207 CEST] [debug] tests/console/consoletest_setup.pm:34 called opensusebasetest::select_serial_terminal -> lib/opensusebasetest.pm:1113 called testapi::select_console -> lib/susedistribution.pm:746 called serial_terminal::login -> lib/serial_terminal.pm:99 called testapi::type_string
[2020-05-12T05:57:33.207 CEST] [debug] <<< testapi::type_string(text="PS1=\"# \"\n")
[2020-05-12T05:57:33.207 CEST] [debug] <<< consoles::serial_screen::type_string(json_cmd_token="hjQGqxgt", text="PS1=\"# \"\n", cmd="backend_type_string")
[2020-05-12T05:57:33.207 CEST] [debug] tests/console/consoletest_setup.pm:34 called opensusebasetest::select_serial_terminal -> lib/opensusebasetest.pm:1113 called testapi::select_console -> lib/susedistribution.pm:746 called serial_terminal::login -> lib/serial_terminal.pm:100 called testapi::wait_serial
```

File: tests/console/ant.pm (hunk 2 above)
Comment (dzedro) [reply tests/console/ant.pm:94]: On the top you will see the content of `$java_hello_world`, it is much better and stable for longer complex text use script_output, it will execute it as script, than (assert_)script_run which will execute it as command and possibly fail.
http://10.100.12.155/tests/15694#step/ant/23

File: tests/console/gdb.pm (hunk 5 above)
Comment (dzedro) [reply tests/console/gdb.pm:44]: Yes, but as I went trough the tests, I was cleaning up. I brushed this not needed code with the rest. I agree and will separate it when you insist. :innocent:

File: tests/installation/bootloader_uefi.pm (hunk 6 above)
Comment (Vogtinator) [reply tests/installation/bootloader_uefi.pm:60]: IMO this needs a separate discussion. @mloviska has a similiar hack in tests/jeos/grub2.pm already, so adding a duplicate here is not great.

File: lib/serial_terminal.pm (hunk 1 above)
Comment (Vogtinator) [reply lib/serial_terminal.pm:101]: I know, but this would just work just fine AFAICT:

```
if (length $testapi::password) {
    ....
    type_string("\n");
}

die 'Failed to confirm that login was successful' unless wait_serial(qr/.*:~\s\#/);
```

File: tests/console/ant.pm (hunk 2 above)
Comment (Vogtinator) [reply tests/console/ant.pm:94]: Seems like `assert_script_run` does it too: http://10.100.12.155/tests/15694#step/ant/17

Looking at the code, `script_output` is more complex than `assert_script_run` and in some cases even uses `assert_script_run`.

It just seems weird to me to use `script_output` for something which doesn't output anything.
If there is a specific issue with `assert_script_run`, that should probably be fixed in os-autoinst instead.

File: lib/serial_terminal.pm (hunk 1 above)
Comment (dzedro) [reply lib/serial_terminal.pm:101]: yes, I will do that.

File: tests/installation/bootloader_uefi.pm (hunk 6 above)
Comment (mloviska) [reply tests/installation/bootloader_uefi.pm:60]: I can see that the community JeOS ( aarch64 ) still uses `bootloader_uefi.pm`. 
Not sure, how to proceed over here. Whether to put the grub2.pm into generic lib or if we want to modify the schedule. WDYT @Vogtinator, @dzedro

Comment (dzedro) [reply tests/installation/bootloader_uefi.pm:60]: I will drop this then and create new PR

Comment (Vogtinator) [reply tests/installation/bootloader_uefi.pm:60]: @mloviska Is there any reason that JeOS can't use the generic bootloader modules? It would IMO be the best option here if we can avoid having a specific `tests/jeos/grub2.pm`.

File: tests/console/ant.pm (hunk 2 above)
Comment (mloviska) [reply tests/console/ant.pm:94]: Too much input for my taste, I would rather prefer to upload the code from `data` folder.

File: lib/serial_terminal.pm (hunk 1 above)
Comment (dzedro) [reply lib/serial_terminal.pm]: done

File: tests/console/ant.pm (hunk 2 above)
Comment (dzedro) [reply tests/console/ant.pm]: dropped ant, will check it out later

File: tests/installation/bootloader_uefi.pm (hunk 6 above)
Comment (dzedro) [reply tests/installation/bootloader_uefi.pm:60]: dropped, I will crate separate PR for it.

File: tests/console/mta.pm [hunk 8]
```diff
@@ -37,7 +38,7 @@ sub run {
     # test email transmission
     assert_script_run 'echo "FOOBAR123" | mail root';
     assert_script_run 'postqueue -p';
-    script_run 'cat /var/mail/root';
+    sleep(1);
```
Comment (foursixnine) [inline tests/console/mta.pm:41]: This is due to postfix not processing/delivering the mail fast enough right?

Comment (dzedro) [reply tests/console/mta.pm:41]: Yes, try to run this command anywhere, I could see on every machine the first mail in queue.

```
# echo "FOOBAR123" | mail root && postqueue -p && grep FOOBAR123 /var/mail/root
Mail queue is empty
FOOBAR123
scruffy:~ # echo "FOOBAR123" | mail root && postqueue -p && grep FOOBAR123 /var/mail/root
Mail queue is empty
FOOBAR123
FOOBAR123
scruffy:~ # echo "FOOBAR123" | mail root && postqueue -p && grep FOOBAR123 /var/mail/root
Mail queue is empty
FOOBAR123
```

Comment (Vogtinator) [reply tests/console/mta.pm:41]: This should be fixed in a better way then and not as part of this commit IMO.

File: tests/console/libqca2.pm [hunk 9]
```diff
@@ -60,8 +60,10 @@ sub run {
     assert_script_run "$qca_cmd show cert cert.pem";
     assert_script_run "$qca_cmd keybundle make rsapriv.pem cert.pem --pass=suse --newpass=suse";
     assert_script_run "$qca_cmd keystore list-stores";
-    script_run "$qca_cmd keystore monitor";
+    type_string "$qca_cmd keystore monitor\n";
+    sleep(1);
```
Comment (Vogtinator) [inline tests/console/libqca2.pm:65]: That looks like it should wait for the prompt to appear instead . Separate PR (or at least commit) please.

File: tests/console/git.pm [hunk 10]
```diff
@@ -22,12 +22,12 @@ use base "consoletest";
...
 use warnings;
 use testapi;
-use utils;
+use utils qw(zypper_call);
 
 sub run {
-
     my $username = $testapi::username;
     my $email    = "you\@example.com";
+    my ($self)   = shift;
```
Comment (Vogtinator) [inline tests/console/git.pm]: Not needed?

Comment (dzedro) [reply tests/console/git.pm]: This was outdated when you commented it.

File: tests/installation/bootloader_uefi.pm (hunk 6 above)
Comment (mloviska) [reply tests/installation/bootloader_uefi.pm]: > @mloviska Is there any reason that JeOS can't use the generic bootloader modules? It would IMO be the best option here if we can avoid having a specific `tests/jeos/grub2.pm`.

Images have reduced timeout, compared to isos at first boot stage. Race condition happens at this point, in case we want to interact with grub to set gfxpayload https://openqa.suse.de/tests/4238030#step/grub2/3. 
But as you already pointed out, it is not necessary to set `gfxpayload` when we use `virtio` driver. 
http://kepler.suse.cz/tests/29#step/console_reboot/11

File: tests/console/mta.pm (hunk 8 above)
Comment (dzedro) [reply tests/console/mta.pm:41]: Can't wait for the improvement on this puzzling issue.
I see only simple and short wait for mail queue to get processed in super fast test.

Comment (Vogtinator) [reply tests/console/mta.pm:41]: It could be changed to `wait_serial` on `tail -f` and send `ctrl-c` afterwards for instance.

File: tests/console/libqca2.pm (hunk 9 above)
Comment (dzedro) [reply tests/console/libqca2.pm]: dropped, can be improved in separate PR

File: tests/console/mta.pm (hunk 8 above)
Comment (dzedro) [reply tests/console/mta.pm]: good enough ?

Comment (foursixnine) [reply tests/console/mta.pm]: I was actually fine with the first version.

A decent fix would be to check the postfix queue until the email is actually delivered (waitserial on the journal?) and then check the mail queue... but this is outside of the scope of this PR FWIW

Comment (dzedro) [reply tests/console/mta.pm]: that's what `assert_script_run 'until postqueue -p|grep "Mail queue is empty";do sleep 1;done';` does or ?

Comment (foursixnine) [reply tests/console/mta.pm]: > that's what `assert_script_run 'until postqueue -p|grep "Mail queue is empty";do sleep 1;done';` does or ?

Yeah :)

File: tests/x11/evolution/evolution_prepare_servers.pm [hunk 11]
```diff
@@ -93,21 +93,9 @@ sub run() {
...
     assert_script_run "useradd -m admin";
-    script_run "passwd admin", 0;    # set user's password
-    type_password "password123";
-    wait_still_screen(1);
-    send_key 'ret';
-    type_password "password123";
-    wait_still_screen(1);
-    send_key 'ret';
-
+    assert_script_run "expect -c 'spawn passwd admin;expect sword:;send password123\\r;expect sword:;send password123\\r;interact'";
```
Comment (Vogtinator) [inline tests/x11/evolution/evolution_prepare_servers.pm:98]: You could just run `echo -e 'admin:password123\nnimda:password123' | chpasswd`

File: tests/console/sshd.pm [hunk 12]
```diff
@@ -135,7 +136,6 @@ sub run {
 
     assert_script_run "killall -u $ssh_testman || true";
     wait_still_screen 3;
-    clear_console;
```
Comment (Vogtinator) [inline tests/console/sshd.pm:138]: I assume this was added so that the `job killed` messages are cleared. If there is a needle matching the prompt, it might not match at this point anymore.

Comment (dzedro) [reply tests/console/sshd.pm:138]: change to `clear_console if !is_serial_terminal;`

File: tests/x11/evolution/evolution_prepare_servers.pm (hunk 11 above)
Comment (dzedro) [reply tests/x11/evolution/evolution_prepare_servers.pm]: changed

**Comment by rwx788** [comment]
Hey! We got this failure here caused by this change: https://openqa.suse.de/tests/4265927#step/zypper_ref/1 
Any hints how to fix it quickly?

**Comment by dzedro** [comment]
> Hey! We got this failure here caused by this change: https://openqa.suse.de/tests/4265927#step/zypper_ref/1
> Any hints how to fix it quickly?

yes, working on it https://progress.opensuse.org/issues/67174

**Comment by rwx788** [comment]
> > Hey! We got this failure here caused by this change: https://openqa.suse.de/tests/4265927#step/zypper_ref/1
> > Any hints how to fix it quickly?
> 
> yes, working on it https://progress.opensuse.org/issues/67174

Great! Thanks a lot!
